### ISSUE TITLE
Segfault while connecting from multiple processes with Postgres 12

### ISSUE THREAD

**Issue by jessebs** [ged/ruby-pg#311]
I'm running Postgres 12.1, Ruby 2.6.5 and pg 1.1.4 on MacOS Catalina.  This also happens on Ruby 2.5.1 and 2.5.7 as well as MacOS Mojave.  It does not happen with Postgres 10 or 11.

Running the following small script will cause a segfault.
```
#!/usr/bin/env ruby
require 'pg'

PG.connect(:host => 'localhost', :user => 'username', :dbname => 'test_db')

Process.fork do
  PG.connect(:host => 'localhost', :user => 'username', :dbname => 'test_db')
end
```

See the attached [segfault.txt](https://github.com/ged/ruby-pg/files/3918326/segfault.txt) and [crash_report.txt](https://github.com/ged/ruby-pg/files/3918377/crash_report.txt)

**Comment by konung** [comment]
Confirming the same issue. Ran into it while testing on my MacOS an app, after upgrading to Postgres 12 from Postgres 11 ( via brew) . Works for a Single connection, but crashes for multiple, i.e. crashes a Rails app, when trying to load models.  

```
gem 'pg' => PG::VERSION = "1.1.4"
rails 6.0.1
psql (PostgreSQL) 12.1
MacOS Catalina
``

**Comment by cbandy** [comment]
@jessebs I expect `gssencmode: "disable"` will avoid the crash. Can you please verify?

**Comment by jessebs** [comment]
Yes, that did avoid the issue.

**Comment by cbandy** [comment]
I've reported this upstream: https://postgr.es/m/93f7379b-2e2f-db0c-980e-07ebd5de92ff%40crunchydata.com

**Comment by trevorcreech** [comment]
It looks like in Rails 5, this can be fixed by adding `gssencmode: disable` as a connection parameter in `database.yml`

Unfortunately Rails 4 doesn't pass `gssencmode` along to `pg` (because `VALID_CONN_PARAMS` [is hardcoded until Rails 5](https://github.com/rails/rails/commit/86a23faa1a82a3e7a2cdf2df01eb6eb18dd3d6f6)). We were able to fix this by setting a `PGGSSENCMODE="disable"` environment variable.

**Comment by bengalamx** [comment]
Thanks a lot @trevorcreech, ENV variable PGGSSENCMODE="disable" worked for me.

**Comment by jessebs** [comment]
I'm a bit confused - based on my understanding of the problem, I would expect that closing the connection before the fork would resolve it but it does not.

Any understanding of why that's the case?

**Comment by cbandy** [comment]
It's not about shared handles or anything. It seems that macOS libraries are generally not safe to use after fork() (due to libdispatch.) It's a [known issue](https://github.com/Homebrew/homebrew-core/pull/47494#issuecomment-562400120) over at Homebrew.

**Comment by alienxp03** [comment]
Another workaround that I found is by not setting any value to the host, or just omit the `host` params.

```ruby
# 1
Process.fork do
  PG.connect(:host => '', :user => 'username', :dbname => 'test_db')
end

# 2
Process.fork do
  PG.connect(:user => 'username', :dbname => 'test_db')
end
```

If you're in Rails, you can just do the same thing with the `host` params.

```ruby
# 1: This will crash
development:
  adapter: postgresql
  database: my_database
  host: localhost

# 2: This won't
development:
  adapter: postgresql
  database: my_database
  host:
```

So far I noticed that I only need this workaround if the host is `localhost` or `127.0.0.1`.

**Comment by cbandy** [comment]
@alienxp03 Yes, this also skips GSS encryption by connecting over Unix socket (which is only available when client and server are on the same machine.) Depending on how the server is configured, Unix socket may involve [different authentication](https://www.postgresql.org/docs/current/auth-pg-hba-conf.html) than TCP socket.

**Comment by TiuTalk** [comment]
Just letting you guys know that this also happens on Ruby 2.3.1, using `pg` 0.19.0, with `postgresql` stable 12.1 from Homebrew.

Adding `gssencmode: disable` to the database.yaml fixed the segfaults

**Comment by ankane** [comment]
If you installed PostgreSQL through Homebrew, reinstalling should fix it, thanks to this PR: Homebrew/homebrew-core#47494

```sh
brew uninstall postgresql
# then kill all postgresql processes or restart machine
brew install postgresql
```

**Comment by abepark01** [comment]
I encountered this issue today on Mac OS Ventura 13.8.

postgresql 14.8 (Homebrew)
ruby 3.2.2 (installed via rbenv)
pg 1.5.3
rails 7.0.5

My workaround was to add  `export PGGSSENCMODE="disable"` to my `.zshrc` file

**Comment by kefimochi** [comment]
Encountered this issue today on Venture@13.2.1. Workaround was also `export PGGSSENCMODE="disable"` to my `.zshrc` file. Thanks everyone!

**Comment by jk779** [comment]
same here, came up just today after an auto-upgrade from brew. maybe a regression?

@ankane's solution did not work for me :( 

so i had to go the route with `PGGSSENCMODE="disable"` too, which i don't really like :/

**Comment by DaniG2k** [comment]
Same issue here. Although not ideal, setting nothing for the `host:` value like the following in `database.yml` seems to do the trick:

```yml
development:
  host:
  adapter: postgresql
  encoding: unicode
  database: <%= ENV['DATABASE_NAME'] %>
  pool: 5
```

Again, this is a temporary solution and by no means ideal.

**Comment by daande** [comment]
@konung it seems people are experiencing this again.  Can we reopen this issue?

**Comment by stanhu** [comment]
This is likely happening because `krb5` v1.21 shipped with this change: https://github.com/krb5/krb5/pull/1221

My colleague shared this relevant backtrace:

```json
  "vmRegionInfo" : "0x103e00abc is not in any region.  Bytes after previous region: 2749  Bytes before following region: 62788\n      REGION TYPE                    START - END         [ VSIZE] PRT\/MAX SHRMOD  REGION DETAIL\n      MALLOC_TINY                 103d00000-103e00000    [ 1024K] rw-\/rwx SM=PRV  \n--->  GAP OF 0x10000 BYTES\n      VM_ALLOCATE                 103e10000-103e20000    [   64K] rw-\/rwx SM=COW  ",
  "exception" : {"codes":"0x0000000000000001, 0x0000000103e00abc","rawCodes":[1,4359981756],"type":"EXC_BAD_ACCESS","signal":"SIGABRT","subtype":"KERN_INVALID_ADDRESS at 0x0000000103e00abc"},
  "vmregioninfo" : "0x103e00abc is not in any region.  Bytes after previous region: 2749  Bytes before following region: 62788\n      REGION TYPE                    START - END         [ VSIZE] PRT\/MAX SHRMOD  REGION DETAIL\n      MALLOC_TINY                 103d00000-103e00000    [ 1024K] rw-\/rwx SM=PRV  \n--->  GAP OF 0x10000 BYTES\n      VM_ALLOCATE                 103e10000-103e20000    [   64K] rw-\/rwx SM=COW  ",
  "asi" : {"CoreFoundation":["*** multi-threaded process forked ***"],"libsystem_c.dylib":["crashed on child side of fork pre-exec"]},
  "extMods" : {"caller":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"system":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"targeted":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"warnings":0},
  "faultingThread" : 0,
  "threads" : [{"triggered":true,"id":139232,"threadState":{"x":[{"value":0},{"value":0},{"value":0},{"value":0},{"value":20680267530240},{"value":4410931412992},{"value":144},{"value":0},{"value":10969316120912989786},{"value":10969316121450844250},{"value":2},{"value":4294967293},{"value":1099511627776},{"value":0},{"value":0},{"value":0},{"value":328},{"value":8131268448},{"value":0},{"value":6},{"value":8052481536,"symbolLocation":0,"symbol":"_main_thread"},{"value":771},{"value":8052481760,"symbolLocation":224,"symbol":"_main_thread"},{"value":4303728177,"symbolLocation":33915,"symbol":"hex_table"},{"value":95},{"value":0},{"value":0},{"value":0},{"value":0}],"flavor":"ARM_THREAD_STATE64","lr":{"value":6525926440},"cpsr":{"value":1073745920},"fp":{"value":6171233216},"sp":{"value":6171233184},"esr":{"value":1442840704,"description":" Address size fault"},"pc":{"value":6525699876,"matchesCrashFrame":1},"far":{"value":4701421568}},"queue":"com.apple.main-thread","frames":[{"imageOffset":38692,"symbol":"__pthread_kill","symbolLocation":8,"imageIndex":63},{"imageOffset":27688,"symbol":"pthread_kill","symbolLocation":288,"imageIndex":64},{"imageOffset":486120,"symbol":"abort","symbolLocation":180,"imageIndex":65},{"imageOffset":556468,"symbol":"die","symbolLocation":12,"imageIndex":1},{"imageOffset":556916,"symbol":"rb_bug_for_fatal_signal","symbolLocation":448,"imageIndex":1},{"imageOffset":1792864,"symbol":"sigsegv","symbolLocation":96,"imageIndex":1},{"imageOffset":14884,"symbol":"_sigtramp","symbolLocation":56,"imageIndex":66},{"imageOffset":18052,"symbol":"_os_log_preferences_refresh","symbolLocation":36,"imageIndex":67},{"imageOffset":20748,"symbol":"os_log_type_enabled","symbolLocation":712,"imageIndex":67},{"imageOffset":44020,"symbol":"_xpc_connection_activate_if_needed","symbolLocation":152,"imageIndex":68},{"imageOffset":54464,"symbol":"xpc_connection_resume","symbolLocation":92,"imageIndex":68},{"imageOffset":51696,"symbol":"get_primary_name","symbolLocation":152,"imageIndex":38},{"imageOffset":50396,"symbol":"api_macos_ptcursor_next","symbolLocation":240,"imageIndex":38},{"imageOffset":38792,"symbol":"krb5_cccol_cursor_next","symbolLocation":76,"imageIndex":38},{"imageOffset":39536,"symbol":"krb5_cccol_have_content","symbolLocation":92,"imageIndex":38},{"imageOffset":88212,"symbol":"acquire_cred_context","symbolLocation":1664,"imageIndex":37},{"imageOffset":86428,"symbol":"acquire_cred_from","symbolLocation":688,"imageIndex":37},{"imageOffset":29056,"symbol":"gss_add_cred_from","symbolLocation":624,"imageIndex":37},{"imageOffset":28104,"symbol":"gss_acquire_cred_from","symbolLocation":400,"imageIndex":37},{"imageOffset":27692,"symbol":"gss_acquire_cred","symbolLocation":36,"imageIndex":37},{"imageOffset":105780,"symbol":"pg_GSS_have_cred_cache","symbolLocation":60,"imageIndex":34},{"imageOffset":34308,"symbol":"PQconnectPoll","symbolLocation":4416,"imageIndex":34},{"imageOffset":10032,"symbol":"gvl_PQconnectPoll_skeleton","symbolLocation":24,"imageIndex":33},{"imageOffset":2034236,"symbol":"rb_nogvl","symbolLocation":268,"imageIndex":1},{"imageOffset":9992,"symbol":"gvl_PQconnectPoll","symbolLocation":44,"imageIndex":33},{"imageOffset":33088,"symbol":"pgconn_connect_poll","symbolLocation":48,"imageIndex":33},{"imageOffset":2361060,"symbol":"vm_call_cfunc_with_frame","symbolLocation":232,"imageIndex":1},{"imageOffset":2363212,"symbol":"vm_call_symbol","symbolLocation":572,"imageIndex":1},{"imageOffset":2252892,"symbol":"vm_exec_core","symbolLocation":8132,"imageIndex":1},{"imageOffset":2325636,"symbol":"rb_vm_exec","symbolLocation":2092,"imageIndex":1},
```

As explained in https://blog.phusion.nl/2017/10/13/why-ruby-app-servers-break-on-macos-high-sierra-and-what-can-be-done-about-it/, this happens because the macOS system calls are not thread-safe, so when they get called in a fork a seg fault may occur.

I suspect there's not much this library can do about it, but there are workarounds:

1. Disable GSSAPI via the `PGGSSENCMODE=disable` environment variable or passing `gssencmode=disable` in the connection string.
2. Preloading the app before forking. In Puma, the `preload_app!` config option does this.

**Comment by Meekohi** [comment]
`PGGSSENCMODE="disable"` is not working for me, is there an earlier version of `pg` that is not affected?

**Comment by jdeff** [comment]
@Meekohi if you are using rails, try adding `gssencmode: disable` to your db config in `config/database.yml`.

**Comment by Meekohi** [comment]
I did give that a shot as well with no luck. After more experiments I found that this was only effecting me on the new Amazon Linux AMI2023, and does *not* effect the older Amazon Linux AMI2, so I've switched back to that for now. I assume there is some newer library (maybe libpq?) the new AMI is using that somehow exercises this bug.

**Comment by Rhoxio** [comment]
> It looks like in Rails 5, this can be fixed by adding `gssencmode: disable` as a connection parameter in `database.yml`
> 
> Unfortunately Rails 4 doesn't pass `gssencmode` along to `pg` (because `VALID_CONN_PARAMS` [is hardcoded until Rails 5](https://github.com/rails/rails/commit/86a23faa1a82a3e7a2cdf2df01eb6eb18dd3d6f6)). We were able to fix this by setting a `PGGSSENCMODE="disable"` environment variable.

This fixed my issue. Thank you so much!

**Comment by stevenharman** [comment]
> 2. Initiate the database connection by preloading the app before forking so the macOS system calls are invoked in the parent process. In Puma, the preload_app! config option does this.

This just does not work for me. I've even tried the following `before_fork` hook in Puma:

```ruby
before_fork do
  require 'pg'
  # Force a connection and query to PG
  ActiveRecord::Base.connection.migration_context.current_version

  # Work around macOS 10.13 and later being very picky about `fork` usage and
  # interactions with Objective-C code see:https://github.com/puma/puma/issues/1421
  if /darwin/ =~ RUBY_PLATFORM
    require 'fiddle'
    # Dynamically load Foundation.framework, ~implicitly~ initialising the
    # Objective-C runtime before any forking happens in Puma
    Fiddle.dlopen('/System/Library/Frameworks/Foundation.framework/Foundation')
  end

  ActiveRecord::Base.connection_pool.disconnect! if defined?(ActiveRecord)
end
```

I can set `gssencmode: disable` in `database.yml`, but that seems so heavy-handed, and preloading the app sure sounds like it should fix it.

Ruby: `3.0.6`
pg: `1.5.4`
Rails: `6.1.7.4`

**Comment by docwhat** [comment]
> `PGGSSENCMODE="disable"` is not working for me, is there an earlier version of `pg` that is not affected?

The actual command is `export PGGSSENCMODE="disable"`. Without `export`, the variable is only visible to the shell, not to any commands (like `rake`, `ruby`, or `rails`).

You probably knew that, but I wanted to clarify for others.

**Comment by stanhu** [comment]
> This just does not work for me. I've even tried the following before_fork hook in Puma:

Is it possible there's a different seg fault happening (such as https://github.com/ged/ruby-pg/issues/555)? I should note that if I modify the reproduction script in https://github.com/ged/ruby-pg/issues/311#issue-532214355 to incorporate your change, this seems to fix the GSSAPI seg fault for me:

```ruby
#!/usr/bin/env ruby
require 'pg'

PG.connect(:host => 'localhost', :user => 'username', :dbname => 'test_db')

if /darwin/ =~ RUBY_PLATFORM
  require 'fiddle'
  # Dynamically load Foundation.framework, ~implicitly~ initialising the
  # Objective-C runtime before any forking happens in Puma
  Fiddle.dlopen('/System/Library/Frameworks/Foundation.framework/Foundation')
end

Process.fork do
  PG.connect(:host => 'localhost', :user => 'username', :dbname => 'test_db')
end

**Comment by napster235** [comment]
> @Meekohi if you are using rails, try adding `gssencmode: disable` to your db config in `config/database.yml`.

this worked for me!

**Comment by PrathameshSurve** [comment]
If you are using postgresql 12, ruby 2.6.5 rails ^6.0.0 , on ubuntu 22.04 and segmentation fault occures 
then this is for you///////////////////////////////->

by default ubuntu 22.04  uses openssl 3 but your ruby version needs openssl 1.0.1 or 1.1.1 so there is nothing t o do with that, ubuntu doen't allow openssl 1.1.1  or 1.0.1 so there are some other options you can go for, 


check the database.yml file and replace host: localhost to host : 
yes just keep host as a blank in your database.yml file. 
before that if your environment doen't allow you to install ruby then you have another problem, for another problems related with versions ping me on personal.

**Comment by BrijeshSajeev** [comment]
I'm running Postgres 12.1, Ruby 2.7.4 and pg 1.4.6 on MX-linux and i'm facing the segfault and i tried gssencmode: disable but still no luck.

**Comment by bichou** [comment]
> I'm running Postgres 12.1, Ruby 2.7.4 and pg 1.4.6 on MX-linux and i'm facing the segfault and i tried gssencmode: disable but still no luck.

try this work for me 
systemctl stop postgresql.service
msfconsole 
work fine

**Comment by Tectract** [comment]
bump, experiencing still in 2025 now. PG should like, work with rails.

**Comment by rromanchuk** [comment]
Here from the future, @harsh183 did you ever figure out why this seems like it regressed recently? Was it specific to 26.2 or did my own behavior change and now just only noticing it.  I'm also aggressively using bin/ci now, or i guess, a lot more forking workers in general. When i first saw this ignored it because of the year, but this is forsure the issue. I'm also able to "fix"  with PARALLEL_WORKERS=1 which is another dead give away.

**Comment by harsh183** [comment]
https://notes.max.engineer/ruby-pg-gem-segfault

best one I've found but it stayed broken across several mac os versions. My projects have quite a few tests and my MacBook has 12 cores so I want to use all of them to crunch so I just disable it. 

At least I've only configured this hack to be local dev and ci. Ownership of fixing this seems so diffused, I'm not sure if someone will handle it. 

I've only seen it on minitest, and I've also run into a lot because of all the bin/ci I use recently
